src/libAtomVM/externalterm.c

```diff
@@ -54,22 +57,217 @@ term externalterm_to_term(const void *external_term, Context *ctx)
     int eterm_size;
     int heap_usage = calculate_heap_usage(external_term_buf + 1, &eterm_size, ctx);
 
-    struct ListHead *heap_fragment = malloc(heap_usage * sizeof(term) + sizeof(struct ListHead));
-    if (IS_NULL_PTR(heap_fragment)) {
+    if (use_heap_fragment) {
+        struct ListHead *heap_fragment = malloc(heap_usage * sizeof(term) + sizeof(struct ListHead));
+        if (IS_NULL_PTR(heap_fragment)) {
+            return term_invalid_term();
+        }
+        list_append(&ctx->heap_fragments, heap_fragment);
+        ctx->heap_fragments_size += heap_usage;
+        term *external_term_heap = (term *) (heap_fragment + 1);
+
+        // save the heap pointer and temporary switch to the newly created heap fragment
+        // so all existing functions can be used on the heap fragment without any change.
+        term *main_heap = ctx->heap_ptr;
+        ctx->heap_ptr = external_term_heap;
+        term result = parse_external_terms(external_term_buf + 1, &eterm_size, ctx);
+        ctx->heap_ptr = main_heap;
+
+        return result;
+    } else {
+        if (UNLIKELY(memory_ensure_free(ctx, heap_usage) != MEMORY_GC_OK)) {
+            fprintf(stderr, "Unable to ensure %i free words in heap\n", eterm_size);
+            return term_invalid_term();
+        }
+        term result = parse_external_terms(external_term_buf + 1, &eterm_size, ctx);
+        return result;
+    }
+}
+
+term externalterm_from_binary(Context *ctx, term binary)
+{
+    if (!term_is_binary(binary)) {
+        return term_invalid_term();
+    }
+    //
+    // Copy the binary data to a buffer (in case of GC)
+    //
+    size_t len = term_binary_size(binary);
+    const uint8_t *data = (const uint8_t *) term_binary_data(binary);
+    uint8_t *buf = malloc(len);
+    if (UNLIKELY(IS_NULL_PTR(buf))) {
+        fprintf(stderr, "Unable to allocate %zu bytes for binary buffer.\n", len);
+        abort();
```

Comment:
We shouldn't abort() anymore, let's raise an out of memory error instead.

Reply:
Fixed.